lib/iterator.js

```diff
@@ -0,0 +1,304 @@
+  }
+
+  skip(amount) {
+    for (let i = 0; i < amount; i++) {
+      this.next();
+    }
+    return this;
+  }
+}
+
+class MapIterator extends Iterator {
+  constructor(base, mapper, thisArg) {
+    super(base);
+    this.mapper = mapper;
+    this.thisArg = thisArg;
+  }
+
+  next() {
+    const { done, value } = this.base.next();
+    return {
+      done,
+      value: done ? undefined : this.mapper.call(this.thisArg, value),
+    };
+  }
+}
+
+class FilterIterator extends Iterator {
+  constructor(base, predicate, thisArg) {
+    super(base);
+    this.predicate = predicate;
+    this.thisArg = thisArg;
+  }
+
+  next() {
+    for (const value of this.base) {
+      if (this.predicate.call(this.thisArg, value)) {
+        return { done: false, value };
+      }
+    }
+    return { done: true, value: undefined };
+  }
+}
+
+class FlatIterator extends Iterator {
+  constructor(base, depth) {
+    super(base);
+    this.currentDepth = 0;
+    this.stack = new Array(depth + 1);
+    this.stack[0] = base;
+  }
+
+  next() {
+    while (this.currentDepth >= 0) {
+      const top = this.stack[this.currentDepth];
+      const next = top.next();
+
+      if (next.done) {
+        this.stack[this.currentDepth] = null;
+        this.currentDepth--;
+        continue;
+      }
+
+      if (
+        this.currentDepth === this.stack.length - 1 ||
+        !next.value[Symbol.iterator]
+      ) {
+        return next;
+      }
+
+      this.stack[++this.currentDepth] = next.value[Symbol.iterator]();
+    }
+
+    return { done: true, value: undefined };
+  }
+}
+
+class FlatMapIterator extends Iterator {
+  constructor(base, mapper, thisArg) {
+    super(base);
+    this.mapper = mapper;
+    this.thisArg = thisArg;
+    this.currentIterator = null;
+  }
+
+  next() {
+    if (!this.currentIterator) {
+      const next = this.base.next();
+      if (next.done) {
+        return next;
+      }
+
+      const value = this.mapper.call(this.thisArg, next.value);
+      if (!value[Symbol.iterator]) {
+        return { done: false, value };
+      }
+
+      this.currentIterator = toIterator(value);
+    }
+
+    const next = this.currentIterator.next();
+
+    if (next.done) {
+      this.currentIterator = null;
+      return this.next();
+    }
+    return next;
+  }
+}
+
+class TakeIterator extends Iterator {
+  constructor(base, amount) {
+    super(base);
+    this.amount = amount;
+    this.iterated = 0;
+  }
+
+  next() {
+    this.iterated++;
+    if (this.iterated <= this.amount) {
+      return this.base.next();
+    }
+    return { done: true, value: undefined };
+  }
+}
+
+class ZipIterator extends Iterator {
+  constructor(base, iterators) {
+    super(base);
+    this.iterators = iterators.map(toIterator);
+  }
+
+  next() {
+    const result = [];
+
+    const next = this.base.next();
+    if (next.done) {
+      return next;
+    }
+    result.push(next.value);
+
+    for (const iterator of this.iterators) {
```

Comment:
In a follow up PR: update this to have `finished` flag to avoid going every time through this loop up to the iterator that is `done`.